libraries/node-core-library/src/PackageJsonLookup.ts

```diff
@@ -100,6 +100,19 @@ export class PackageJsonLookup {
     );
   }
 
+  public static getOwnPackageJsonVersion(dirnameOfCaller: string): string {
+    let parent: string = path.dirname(dirnameOfCaller);
+    do {
+      try {
+        return require(path.resolve(dirnameOfCaller, 'package.json')).version;
```

Comment:
Why isn't this using the same mechanism as a normal `package.json` lookup?

Reply:
The previously used API is a few hundred milliseconds slower than this new method.

Reply:
@sachinjoseph This doesn't seem like a good answer. `PackageJsonLookup` is a heavily used API, so if something is slow about it, we should go fix that.  

Without altering performance, `getOwnPackageJsonVersion()` could just as easily return the whole **package.json** object, making its functionality equivalent to `loadOwnPackageJson()`.  Thus this solution seems like it's adding more code to the pile without fixing the problem.

Also it introduces some potential problems of its own:
- The `catch` block assumes the error is "file not found", which is not always the case
- The `while` loop is potentially infinite because its exit criteria is different from `_tryGetPackageFolderFor()`
- We are calling `require()` which executes JavaScript code.  It is not a proper JSON parser.
- `getOwnPackageJsonVersion()` is not cached -- it will read the disk every time it is called.

Reply:
@octogonz I think we should implement such a function that is more performant to be used by safe code paths such as the callers of this method. If we can make some assumptions - such as errors mean file not found and that a package.json, will be at max at a height of, say 6 levels or so up, then we can come up with a performant yet relatively safe implementation, to be used by trusted code paths.

Reply:
🤔 I wonder if most of the speedup actually comes from calling `require()` instead of `JsonFile.load()`.  If so, maybe `PackageJsonLookup` could expose an API option that enables this optimization (while clearly documenting the limitations it imposes).

Reply:
> we should implement such a function that is more performant to be used by safe code paths such as the callers of this method. If we can make some assumptions - such as errors mean file not found and that a package.json, will be at max at a height of, say 6 levels or so up, then we can come up with a performant yet relatively safe implementation, to be used by trusted code paths.

@sachinjoseph Thinking about this some more, the restrictions you've made don't really impact performance.  You are still loading the entire file, so your algorithm can still return the full `IPackageJson` (but maybe cloning it or freezing it first?).  Otherwise the API signatures and intended usage of both APIs should be the same.

Thus I'd suggest two possible solutions:

1. Optimize `loadOwnPackageJson()` by having it use your algorithm instead of `PackageJsonLookup`, but fix the problem where we are doing disk reads if it isn't called.

2. Even better, add a `IPackageJsonLookupParameters.parseUsingRequire` that optionally uses `require()` instead of `JsonFile.load()`, with some docs explaining the downsides of `require()` that I mentioned above.  If `require()` throws an exception, maybe we could also fallback to `JsonFile.load()` to obtain a better error message.

3. Take it a step further and maybe we call the option `IPackageJsonLookupParameters.parseUsingEval` and do this instead, which might be even faster than `require()`, and would eliminate the caching problem:
    ```ts
    const jsonString = fs.readFileSync('package.json').toString();
    const json = eval('(' + jsonString + ')');
    ```
I feel like #<!-- -->2 or #<!-- -->3 should make the regular `PackageJsonLookup` algorithm as fast as yours, which would would be cool since some other tools might want that as well.

@iclanton

Reply:
I started to prototype this optimization myself, but when I measured the timings for `getOwnPackageJsonVersion()` and `loadOwnPackageJsonVersion()` they are both <10ms.

Reply:
I got similar results. I removed the new API.